File: lcof/of003/Solution.py
from typing import List


class Solution:
    def findRepeatNumber(self, nums: List[int]) -> int:
        dic = set()
        for v in nums:
            if v in dic: return v
            dic.add(v)
        return -1


class Solution2:
    def findRepeatNumber(self, nums: List[int]) -> int:
        nums.sort()
        for idx in range(len(nums) - 1):
            if nums[idx] == nums[idx + 1]:
                return nums[idx]


class Solution3:
    def findRepeatNumber(self, nums: List[int]) -> int:
        for idx, v in enumerate(nums):
            if idx != v and nums[v] == v: return v
            nums[v], nums[idx] = nums[idx], nums[v]


if __name__ == '__main__':
    print(Solution().findRepeatNumber([2, 3, 1, 0, 2, 5, 3]))
    print(Solution2().findRepeatNumber([2, 3, 1, 0, 2, 5, 3]))
    print(Solution3().findRepeatNumber([2, 3, 1, 0, 2, 5, 3]))

File: lcof/of003/Solution.go
package Solution

import "sort"

//	排序 + 遍历
func findRepeatNumber(nums []int) int {
	sort.Ints(nums)
	for idx := range nums {
		if idx > 0 {
			if nums[idx] == nums[idx-1] {
				return nums[idx]
			}
		}
	}
	return 0
}

//	map哈希表
func findRepeatNumber2(nums []int) int {
	mapTmp := make(map[int]bool)
	for _, val := range nums {
		if mapTmp[val] {
			return val
		} else {
			mapTmp[val] = true
		}
	}
	return 0
}

//	原地置换
func findRepeatNumber3(nums []int) int {
	for idx, val := range nums {
		if idx == val {
			continue
		}
		if nums[val] == val {
			return val
		}
		nums[val], nums[idx] = nums[idx], nums[val]
	}
	return 0
}
